OpenRA.Mods.Common/Traits/BotModules/McvExpansionManagerBotModule.cs

```diff
@@ -0,0 +1,822 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public enum BotMcvExpansionMode { CheckResource, CheckBase, CheckCurrentLocation }
+
+	[TraitLocation(SystemActors.Player)]
+	[Desc("Manages AI MCVs and expansion.")]
+	public class McvExpansionManagerBotModuleInfo : ConditionalTraitInfo, Requires<ResourceMapBotModuleInfo>, NotBefore<ResourceMapBotModuleInfo>
+	{
+		[Desc("Actor types that are considered MCVs (deploy into base builders).")]
+		public readonly HashSet<string> McvTypes = [];
+
+		[Desc("Actor types that are considered construction yards (base builders).")]
+		public readonly HashSet<string> ConstructionYardTypes = [];
+
+		[Desc("Actor types that are able to produce MCVs.")]
+		public readonly HashSet<string> McvFactoryTypes = [];
+
+		[Desc("Try to maintain at least this many ConstructionYardTypes, build an MCV if number is below this.")]
+		public readonly int MinimumConstructionYardCount = 1;
+
+		[Desc("Try to maintain at additional this many ConstructionYardTypes.")]
+		public readonly int AdditionalConstructionYardCount = 0;
+
+		[Desc("Build additional MCV if cash is above this.")]
+		public readonly int BuildAdditionalMCVCashAmount = 5000;
+
+		[Desc("Delay (in ticks) between looking for and giving out orders to new MCVs.")]
+		public readonly int ScanForNewMcvInterval = 20;
+
+		[Desc("Delay (in ticks) between check and build a MCV.")]
+		public readonly int BuildMcvInterval = 101;
+
+		[Desc("Move a conyard if have more than 1 conyard, for better expansion.")]
+		public readonly int MoveConyardTick = 5000;
+
+		[Desc("if have more than 1 conyard, move old conyard first.")]
```

Comment:
```suggestion
		[Desc("Should moving the oldest or newest conyard be preferred? Random ordering if unset.")]
```

Granted, allowing to use a null value for a flag is confusing